src/utils/image-processor.js

```diff
@@ -91,90 +92,73 @@ const shrinkFile = function(file, settings) {
 }
 
 const shrinkImage = function(img, settings) {
-  var cx,
-    df,
-    h,
-    maxSize,
-    maxSquare,
-    originalW,
-    ratio,
-    run,
-    runNative,
-    sH,
-    sW,
-    step,
-    w
   // in -> image
   // out <- canvas
-  df = $.Deferred()
-  step = 0.71 // sohuld be > sqrt(0.5)
-  if (img.width * step * img.height * step < settings.size) {
-    return df.reject('not required')
-  }
-  sW = originalW = img.width
-  sH = img.height
-  ratio = sW / sH
-  w = Math.floor(Math.sqrt(settings.size * ratio))
-  h = Math.floor(settings.size / Math.sqrt(settings.size * ratio))
-  maxSquare = 5000000 // ios max canvas square
-  maxSize = 4096 // ie max canvas dimensions
-  run = function() {
-    if (sW <= w) {
-      df.resolve(img)
-      return
+  const df = $.Deferred()
+
+  const maxSizePromise = maxCanvasSize()
+  maxSizePromise.done(({ maxSize, maxSquare }) => {
+    const step = 0.71 // sohuld be > sqrt(0.5)
+    if (img.width * step * img.height * step < settings.size) {
+      return df.reject('not required')
     }
-    return defer(function() {
-      var canvas
-      sW = Math.round(sW * step)
-      sH = Math.round(sH * step)
-      if (sW * step < w) {
-        sW = w
-        sH = h
-      }
-      if (sW * sH > maxSquare) {
-        sW = Math.floor(Math.sqrt(maxSquare * ratio))
-        sH = Math.floor(maxSquare / Math.sqrt(maxSquare * ratio))
-      }
-      if (sW > maxSize) {
-        sW = maxSize
-        sH = Math.round(sW / ratio)
-      }
-      if (sH > maxSize) {
-        sH = maxSize
-        sW = Math.round(ratio * sH)
+    const originalW = img.width
+    let sW = originalW
+    let sH = img.height
+    const ratio = sW / sH
+    const w = Math.floor(Math.sqrt(settings.size * ratio))
+    const h = Math.floor(settings.size / Math.sqrt(settings.size * ratio))
+
+    const run = function() {
+      if (sW <= w) {
+        df.resolve(img)
+        return
       }
-      canvas = document.createElement('canvas')
-      canvas.width = sW
-      canvas.height = sH
-      canvas.getContext('2d').drawImage(img, 0, 0, sW, sH)
-      img.src = '//:0' // for image
-      img.width = img.height = 1 // for canvas
-      img = canvas
-      df.notify((originalW - sW) / (originalW - w))
-      return run()
-    })
-  }
+      return defer(function() {
+        sW = Math.round(sW * step)
+        sH = Math.round(sH * step)
+        if (sW * step < w) {
+          sW = w
+          sH = h
+        }
+        if (sW * sH > maxSquare) {
+          sW = Math.floor(Math.sqrt(maxSquare * ratio))
+          sH = Math.floor(maxSquare / Math.sqrt(maxSquare * ratio))
+        }
```

Comment:
It's not clear why here is this check. `h` and `h` should already be smaller `maxSquare` and we only reduce this values on each step. Should be investigate can we remove this.